### PR TITLE
Raise an error when lightning replaces an existing sampler

### LINKED ISSUE #1995: Log a warning/ raise an error when lightning replaces an existing Sampler
## 🚀 Feature

First of all, thanks for this awesome project! Really enjoying using it!

Feature Request: Log a warning or raise a MisconfigurationException when lightning replaces an existing sampler with DistributedSampler. 

Even though this behaviour is documented, it's not intuitive. Also, if someone has defined a sampler - it will lead to very different training results if you use Single GPU vs DDP. So best to warn a user to remove the sampler from their code if using `replace_sampler_ddp=True`.

### Motivation

I'm pretty new to lightning - and recently moved one of my models to lightning. Unaware of the details, I chose ddp training strategy (since it was recommended). For a long time, my model was not converging on the val set - and it took me a lot of time to realise what was happening - lightning had replaced my Dataloader's sampler with DistributedSampler! In my particular model, balanced sampling iskey to convergence (since data is highly imbalanced). Lightning's behaviour makes sense in hindsight, and is well documented - but  may have saved me a lot of debugging time had it given a simple warning or error when this happens.

### Additional context

The code change is very trivial. We can take advantage of the fact that DataLoader's default sampler is `SequentialSampler` (if shuffle=False and sampler=None). 

data_loading.py:115
```
if self.replace_sampler_ddp and need_dist_sampler:
    if not isinstance(dataloader.dataset.sampler, SequentialSampler):
        raise MisconfigurationException(...)
```

Does this make sense? Happy to submit a PR if it does.

On a separate note, can anyone here explain why exactly is `DistributedSampler` needed when using DDP? I can see that it will help if I have shuffle=false. It will probably help maintain sanity of an epoch (since otherwise epoch length will become num_processes*epoch).

But let's say I am doing random shuffling - and am cognizant of the fact that my epoch sizes will be higher, there is no real reason that I need to use `DistributedSampler` for DDP right? Or am I missing something?

### PR DESCRIPTION
**Pull request by devashishshankar**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Fixes #1995 

Currently, Trainer replaces the existing sampler with DistributedSampler
if running distributing training and `replace_sampler_ddp=True` (default
behaviour). If a user has configured an existing sampler, this would
lead to widely different results if running a distributed vs
non-distributed training.

This PR fixes this by raising an Error if user has configured a sampler
and uses `replace_sampler_ddp=True`. The recommended behavior from now
on is to either remove the sampler or set `replace_sampler_ddp=False`

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by devashishshankar** [comment]
@Borda I ran into a couple of issues with the build.

Issue 1: Some integration tests (test_gpu.py::test_ddp_all_dataloaders_passed_to_fit, test_horovod.py::test_horovod_cpu, test_horovod.py::test_horovod_cpu_implicit) were failing. This was happening because `EvalModelTemplate` by default configures a train DataLoader with shuffle on ([link](https://github.com/PyTorchLightning/pytorch-lightning/blob/fdbbe968256f6c68a5dbb840a2004b77a618ef61/tests/base/model_utilities.py#L16)). Now, when the tests are running (in distributed mode), this DataLoader is being replaced by a new DataLoader with shuffling off. My change prevents this exact behaviour.

I see three ways to fix this:

1. Pass `replace_sampler_ddp=False` when running the above tests. This is what I implemented to pass the build. However, this has the disadvantage that `auto_add_sampler` part is not getting tested while these integration tests are being run (Unit tests for `auto_add_sampler` still cover this method though.)
2. Change `EvalModelTemplate` to always have shuffle off. Don't know the implications of this.
3. Have two testing methods in `test_gpu` and `test_horovod` (or pytest.parameterize). One which is passed an  `EvalModelTemplate` with shuffle off, and one with on. This may have some implications of build time? And may not be a very trivial change.

Issue 2: Cyclomatic complexity of `auto_add_sampler` had become too high because of the if condition I added. I've fixed it by doing a minor refactoring, by adding a `get_distributed_sampler` inner method. 

Please let me know how you want me to proceed with these issues.

**Comment by devashishshankar** [comment]
@Borda I didn't realize that `DistributedSampler` by default has shuffle on! So actually my condition only makes sense when someone is using a custom sampler (other than SequentialSampler/ RandomSampler). I think that's a neater way to solve the problem and doesn't require any change to existing tests. Updated PR with the same.

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -113,35 +113,44 @@ def auto_add_sampler(self, dataloader: DataLoader, train: bool) -> DataLoader:
...
-                    'ddp_cpu': self.num_processes * self.num_nodes
-                }
-                sampler = DistributedSampler(
-                    dataloader.dataset,
-                    num_replicas=world_size[self.distributed_backend],
-                    rank=self.proc_rank,
-                )
-
-            dl_args['sampler'] = sampler
+            def get_distributed_sampler(dataloader):
```
Comment (devashishshankar) [inline pytorch_lightning/trainer/data_loading.py:129]: This refactoring is unrelated to PR. Done only because the build complained that the function complexity was becoming high (because of the if condition added above)

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:129]: Even though it bypasses the warning message about the complexity, it completely misses the point. 
By wrapping the code into this inner function the code gets even more complex. 
What the warning tried to tell you is move it into a separate function. 
Let's do that instead? :)

Comment (devashishshankar) [reply pytorch_lightning/trainer/data_loading.py:129]: Sure, you want me to make it a method in `TrainerDataLoadingMixin`? I'll make it a private method?
The reason I made it an inner function was because only this function seemed to be using it. But agree that what you're suggesting is more readable :)

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py]: yes very reasonable choice in my opinion. thanks.

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -113,39 +113,48 @@ def auto_add_sampler(self, dataloader: DataLoader, train: bool) -> DataLoader:
         need_dist_sampler = (self.use_ddp or self.use_ddp2 or self.use_horovod or self.use_tpu)
 
         if self.replace_sampler_ddp and need_dist_sampler:
+            if not isinstance(dataloader.sampler, (SequentialSampler, RandomSampler)):
```
Comment (justusschock) [inline pytorch_lightning/trainer/data_loading.py:116]: this way you also have to include `SubsetRandomSampler` and `WeightedRandomSampler` I guess

Comment (devashishshankar) [reply pytorch_lightning/trainer/data_loading.py:116]: Hmm, shouldn't a warning really be raised if one of these is getting replaced by DistributedSampler? The original motivation of this issue was that I had a `WeightedRandomSampler` - and lightning replaced it leading to val acc drop. Took me a long time to debug this. Hence thought that it may be a good idea to raise an Error when an existing sampler is replaced.

Also, if someone uses one of these sampler - won't they get different results if using distributed vs non-distributed training?

Comment (justusschock) [reply pytorch_lightning/trainer/data_loading.py:116]: Maybe you're right about that :)

File: pytorch_lightning/trainer/data_loading.py [hunk 3]
```diff
@@ -113,39 +113,48 @@ def auto_add_sampler(self, dataloader: DataLoader, train: bool) -> DataLoader:
...
+            world_size = {
+                'ddp': self.num_nodes * self.num_processes,
+                'ddp2': self.num_nodes,
+                'ddp_cpu': self.num_processes * self.num_nodes
+            }
+            sampler = DistributedSampler(
+                dataloader.dataset,
+                num_replicas=world_size[self.distributed_backend],
+                rank=self.proc_rank,
+            )
```
Comment (Borda) [inline pytorch_lightning/trainer/data_loading.py:155]:
```suggestion
        if self.use_tpu:
            kwargs = dict(num_replicas=xm.xrt_world_size(), rank=xm.get_ordinal())
        elif self.use_horovod:
            kwargs = dict(num_replicas=hvd.size(), rank=hvd.rank())
        else:
            world_size = {
                'ddp': self.num_nodes * self.num_processes,
                'ddp2': self.num_nodes,
                'ddp_cpu': self.num_processes * self.num_nodes
            }
            kwargs = dict(num_replicas=world_size[self.distributed_backend], rank=self.proc_rank)
        sampler = DistributedSampler(dataloader.dataset, **kwargs)
```

File: tests/trainer/test_dataloaders.py [hunk 4]
```diff
@@ -393,6 +393,21 @@ class CustomDummyObj:
...
+    assert hasattr(result, 'dummy_kwarg')
+
+    class CustomSampler(torch.utils.data.Sampler):
+        pass
+
+    # Should raise an error if existing sampler is being replaced
+    with pytest.raises(MisconfigurationException, match='DistributedSampler'):
+        trainer.auto_add_sampler(
+            CustomDataLoader(list(range(1000)), sampler=CustomSampler(list(range(1000)))),
+            train=True)
```
Comment (Borda) [inline tests/trainer/test_dataloaders.py:409]:
```suggestion
            CustomDataLoader(list(range(1000)), sampler=CustomSampler(list(range(1000)))), train=True)
```
